src/optimizer/levenberg_marquardt_optimizer.rs

```diff
@@ -0,0 +1,175 @@
+
+        let mut last_err: f64 = 1.0;
+
+        for i in 0..opt_option.max_iteration {
+            let (residuals, mut jac) = problem.compute_residual_and_jacobian(&params);
+
+            if i == 0 {
+                // On the first iteration, generate the diagonal of the jacobian.
+                let cols = jac.shape().1;
+                let jacobi_scaling_vec: Vec<(usize, usize, f64)> = (0..cols)
+                    .map(|c| {
+                        let v = jac
+                            .values_of_col(c)
+                            .iter()
+                            .map(|&i| i * i)
+                            .sum::<f64>()
+                            .sqrt();
+                        (c, c, 1.0 / (1.0 + v))
+                    })
+                    .collect();
+
+                jacobi_scaling_diagonal = Some(
+                    faer::sparse::SparseColMat::<usize, f64>::try_new_from_triplets(
+                        cols,
+                        cols,
+                        &jacobi_scaling_vec,
+                    )
+                    .unwrap(),
+                );
+            }
+
+            let current_error = residuals.norm_l2();
+            trace!("iter:{} total err:{}", i, current_error);
+
+            if current_error < opt_option.min_error_threshold {
+                trace!("error too low");
+                break;
+            } else if current_error.is_nan() {
+                log::debug!("solve ax=b failed, current error is nan");
+                return None;
+            }
+            if i > 0 {
+                if (last_err - current_error).abs() < opt_option.min_abs_error_decrease_threshold {
+                    trace!("absolute error decrease low");
+                    break;
+                } else if (last_err - current_error).abs() / last_err
+                    < opt_option.min_rel_error_decrease_threshold
+                {
+                    trace!("relative error decrease low");
+                    break;
+                }
+            }
+            last_err = current_error;
+
+            // Scale the current jacobian by the diagonal matrix
+            jac = jac * jacobi_scaling_diagonal.as_ref().unwrap();
+
+            // J^T * J = Matrix of shape (total_variable_dimension, total_variable_dimension)
+            let jtj = jac
+                .as_ref()
+                .transpose()
+                .to_col_major()
+                .unwrap()
+                .mul(jac.as_ref());
+
+            // J^T * -r = Matrix of shape (total_variable_dimension, 1)
+            let jtr = jac.as_ref().transpose().mul(-&residuals);
+
+            // Regularize the diagonal of jtj between MIN_DIAGONAL and MAX_DIAGONAL.
+            let mut jtj_regularized = jtj.clone();
+            for i in 0..problem.total_variable_dimension {
+                jtj_regularized[(i, i)] =
+                    (jtj.get(i, i).unwrap().max(MIN_DIAGONAL)).min(MAX_DIAGONAL);
```

Comment:
The shape of the matrix is always the same. Is there an unsafe get mut to accelerate this part?

Reply:
Changed it to `[(i, i)]`, which just panics if the access is out of bounds.